docs/madr/decisions/046-meshservice-policy-matching.md

```diff
@@ -0,0 +1,228 @@
+# Policy matching with MeshService
+
+* Status: accepted
+
+Technical Story: #9707
+
+## Context and Problem Statement
+
+Now that MeshService exists, we need to enable users to apply policies in a way that works
+with the new `MeshService` resource. This includes using `StatefulSet`.
+Additionally, policy should be appliable to `MeshMultiZoneService`.
+The inspect API should also reflect policies that use `MeshService`
+which will be covered in a separate MADR.
+
+Some things to remember:
+
+* Currently `kind: MeshService` in `from[].targetRef` is deprecated.
+* While `kind: MeshService` in `spec.targetRef` points to both a Dataplane and a set of inbounds,
+  it is not a canonical reference.
+  Its effects effectively mix with any policies applied
+  to any `MeshService` that points to this `Dataplane`/inbound.
+  A canonical reference can only be a direct choice of an entire `Dataplane` object
+  and potentially a subset of inbounds.
+* A `MeshService` only has a 1-1 relationship when used as `to[].targetRef.kind`
+  since then it corresponds to a particular Envoy outbound listener/some number of clusters.
+
+## Decision Drivers
+
+* Targeting headless `MeshService`
+* Targeting `MeshMultiZoneService`
+* A balance between correctness and ease of use when configuring policy.
+
+## Considered Options
+
+This MADR only consider references in `to[].targetRef.kind` and MeshHTTPRoute/MeshTCPRoute `backendRefs[].kind`.
+
+This MADR adds 1 new kind, `MeshMultiZoneService` and adds support
+for matching real `MeshService` resources to the existing `MeshService` kind.
+Matches on real `MeshService` objects have priority over `kuma.io/service`
+matches.
+
+A majority of the use cases for policy will involve selecting a specific *port* of the
+`MeshService`, `MeshServices` from a given Kuma zone and namespace.
+
+Note, we have already in targetRef:
+
+```yaml
+name: ...
+namespace: ...
+```
+
+Now we have to consider how we attach to a port and select from a zone.
+
+* Add top level field `zone` and `port` and `labels` for selecting groups of
+  `MeshService`
+* Add top level `port` field and use `labels: {"kuma.io/zone": "..."}`
+* Use generic Gateway API targetRef field `sectionName` to specify a port in
+  `targetRef` but `port` in `backendRef` and add `labels` for selecting
+  groups of `MeshService`.
+
+## Decision Outcome
+
+We choose:
+
+* use generic Gateway API targetRef field `sectionName` to specify `port` and
+  add `labels` but use `port` in `backendRefs`.
+
+### `targetRef`
+
+The policy is applied
+on both the Envoy outbound listeners and clusters
+that correspond to the `MeshService`.
+
+References for `kind: MeshService` have the following structure:
+
+```yaml
+---
+kind: MeshService
+spec:
+  ports:
+  - port: 80
+    name: http
+---
+targetRef:
+  kind: MeshService
+  namespace: backend
+  name: backend # only name or labels
+  labels: # only name or labels
+    kuma.io/zone: east
+  sectionName: http
```

Comment:
this field `sectionName` feels weird. I understand that this is taken from Gateway API, but It does not say anything to me. Like, which section are we talking about? Can it be something else then port in the future?  Why can't we just use port like in `backendRef`. Do we already use `sectionName` in other places? 

In my opinion, it adds complexity for regular users, as they need to translate this to port in their head every time they use it

Reply:
No, we don't use `sectionName` yet.

> In my opinion, it adds complexity for regular users, as they need to translate this to port in their head every time they use it.

I don't disagree with this exact point, but if we ever want to be Gateway API first, it adds more complexity just to do things differently.

The main advantage is that we can reuse the same field if we have similar "subsections" in other resources we want to target.

Reply:
so maybe we should use `sectionName` in `backendRef` instead of `port` to make it consistent? Is it the same story here, that GatewayAPI uses `port` field?

Reply:
yeah it does :frowning:

Reply:
it's unfortunate that the api is not
```
section:
  type: Port
  name: http
```

I'm not happy, but I'm ok with `sectionName`, it is confusing without reading docs, but I see value in being consistent with GAPI.

Reply:
@lahabana would like to hear your opinion before we merge

I agree the decision to have just `port: int` in `backendRef` isn't ideal from an API coherency perspective.
At least this way we keep compatibility with Gateway API but leave our options open to maybe add `sectionName` to `backendRef` at a later point.

Reply:
I see the value in being consistent with Gateway API, so I am ok with this api.

Reply:
But port's name in MeshService is not required right? What happens if there is no name?

Reply:
I'd also like us to be consistent with GAPI

Reply:
Can I have examples for both backendRef and targetRef in each case? I tend to lean towards GAPI format but I'm not sure I understand everything